_posts/2020-07-20-fnn-lstm/timeseries-prediction-with-fnn-lstm.Rmd

```diff
@@ -0,0 +1,829 @@
+    rename_with(function(name)
+      paste0(name, "_true")) %>%
+    bind_cols(data.frame(prediction[, , 1] %>% as.array()) %>%
+                rename_with(function(name)
+                  paste0(name, "_pred")))
+  mse <- purrr::map(1:dim(prediction)[2],
+                        function(varno)
+                          calc_mse(comp_df,
+                                   paste0("X", varno, "_true"),
+                                   paste0("X", varno, "_pred"))) %>%
+    unlist()
+  
+  mse
+}
+
+prediction_fnn <- decoder(encoder(test_batch[[1]]))
+mse_fnn <- get_mse(test_batch, prediction_fnn)
+
+prediction_lstm <- model %>% predict(ds_test)
+mse_lstm <- get_mse(test_batch, prediction_lstm)
+
+mses <- data.frame(timestep = 1:n_timesteps, fnn = mse_fnn, lstm = mse_lstm) %>%
+  gather(key = "type", value = "mse", -timestep)
+ggplot(mses, aes(timestep, mse, color = type)) +
+  geom_point() +
+  scale_color_manual(values = c("#00008B", "#3CB371")) +
+  theme_classic() +
+  theme(legend.position = "none") 
+```
+
+And here is the actual comparison. One thing especially jumps to the eye: FNN-LSTM forecast error is significantly lower for
+initial timesteps, first and foremost, for the very first prediction, which from this graph we expect to be pretty good!
+
+```{r, eval=TRUE, echo=FALSE, layout="l-body-outset", fig.cap = "Per-timestep prediction error as obtained by FNN-LSTM and a vanilla stacked LSTM. Green: LSTM. Blue: FNN-LSTM."}
+knitr::include_graphics("images/geyser_mses.png")
+```
+
+
+Interestingly, we see "jumps" in prediction error, for FNN-LSTM, between the very first forecast and the second, and then
+between the second and the ensuing ones, reminding of the similar jumps in variable importance for the latent code! After the
+first ten timesteps, vanilla LSTM has caught up with FNN-LSTM, and we won't interpret further development of the losses based
+on just a single run's output.
+
+Instead, let's inspect actual predictions. We randomly pick sequences from the test set, and ask both FNN-LSTM and vanilla
+LSTM for a forecast. The same procedure will be followed for the other datasets.
+
+```{r, eval = FALSE}
+given <- data.frame(as.array(tf$concat(list(
+  test_batch[[1]][, , 1], test_batch[[2]][, , 1]
+),
+axis = 1L)) %>% t()) %>%
+  add_column(type = "given") %>%
+  add_column(num = 1:(2 * n_timesteps))
+
+fnn <- data.frame(as.array(prediction_fnn[, , 1]) %>%
+                    t()) %>%
+  add_column(type = "fnn") %>%
+  add_column(num = (n_timesteps  + 1):(2 * n_timesteps))
+
+lstm <- data.frame(as.array(prediction_lstm[, , 1]) %>%
+                     t()) %>%
+  add_column(type = "lstm") %>%
+  add_column(num = (n_timesteps + 1):(2 * n_timesteps))
+
+compare_preds_df <- bind_rows(given, lstm, fnn)
+
+plots <- purrr::map(sample(1:dim(compare_preds_df)[2], 16),
+                    function(v)
+                      ggplot(compare_preds_df, aes(num, .data[[paste0("X", v)]], color = type)) +
```

Comment:
I think adding the `{` for functions makes it easier to read.

Reply:
yeah, agree, will do!